File: app/src/main/java/com/yeraygarcia/recipes/database/repository/RecipeDetailRepository.java
package com.yeraygarcia.recipes.database.repository;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.os.AsyncTask;
import android.support.annotation.NonNull;

import com.yeraygarcia.recipes.OnWebResponseListener;
import com.yeraygarcia.recipes.database.AppDatabase;
import com.yeraygarcia.recipes.database.dao.RecipeIngredientDao;
import com.yeraygarcia.recipes.database.dao.ShoppingListDao;
import com.yeraygarcia.recipes.database.entity.Recipe;
import com.yeraygarcia.recipes.database.entity.RecipeIngredient;
import com.yeraygarcia.recipes.database.entity.RecipeStep;
import com.yeraygarcia.recipes.database.entity.ShoppingListItem;
import com.yeraygarcia.recipes.database.entity.custom.UiRecipeIngredient;
import com.yeraygarcia.recipes.database.remote.Request;
import com.yeraygarcia.recipes.database.remote.ResourceData;
import com.yeraygarcia.recipes.database.remote.RetrofitClient;
import com.yeraygarcia.recipes.database.remote.Webservice;
import com.yeraygarcia.recipes.util.Debug;
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModel;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.UUID;

import retrofit2.Call;

public class RecipeDetailRepository {

    private Context mContext;
    private AppDatabase mDb;
    private Webservice mWebservice;

    public RecipeDetailRepository(Application application) {
        mContext = application;
        mDb = AppDatabase.getDatabase(application);
        mWebservice = RetrofitClient.get(mContext).create(Webservice.class);
    }

    public LiveData<Recipe> getRecipeById(UUID id) {
        return mDb.getRecipeDao().findById(id);
    }

    public LiveData<List<UiRecipeIngredient>> getIngredientsByRecipeId(UUID recipeId) {
        return mDb.getRecipeIngredientDao().findByRecipeId(recipeId);
    }

    public LiveData<List<RecipeStep>> getStepsByRecipeId(UUID recipeId) {
        return mDb.getRecipeStepDao().findByRecipeId(recipeId);
    }

    public void update(Recipe recipe) {
        new Request<Recipe>(mContext) {
            @Override
            public Recipe getEntityBeforeUpdate(Recipe newEntity) {
                return mDb.getRecipeDao().findRawById(newEntity.getId());
            }

            @Override
            public void updateLocalDatabase(Recipe entity) {
                mDb.getRecipeDao().update(entity);
            }

            @NotNull
            @Override
            public Call<ResourceData<Recipe>> sendRequestToServer(Recipe newEntity) {
                return mWebservice.updateRecipe(newEntity);
            }

            @Override
            public void onSuccess(Recipe result) {
                // do nothing
            }

            @Override
            public void onError(@NonNull String errorCode, @NonNull String errorMessage) {
                // do nothing
            }
        }.send(recipe);
    }

    public void persistDraft(Recipe recipe, RecipeDetailViewModel recipeDetailViewModel) {
        new Request<Recipe>(mContext) {
            @Override
            public Recipe getEntityBeforeUpdate(Recipe newEntity) {
                return mDb.getRecipeDao().findRawById(newEntity.getId());
            }

            @Override
            public void updateLocalDatabase(Recipe entity) {
                mDb.getRecipeDao().update(entity);
            }

            @NotNull
            @Override
            public Call<ResourceData<Recipe>> sendRequestToServer(Recipe newEntity) {
                return mWebservice.updateRecipe(newEntity);
            }

            @Override
            public void onSuccess(Recipe result) {
                recipeDetailViewModel.setRecipeDraft(null);
            }

            @Override
            public void onError(@NonNull String errorCode, @NonNull String errorMessage) {
                // do nothing
            }
        }.send(recipe);
    }

    public LiveData<Boolean> isInShoppingList(UUID recipeId) {
        return mDb.getShoppingListDao().isInShoppingList(recipeId);
    }

    public void removeFromShoppingList(UUID recipeId) {
        new RemoveFromShoppingListAsyncTask(mDb.getShoppingListDao()).execute(recipeId);
    }

    public void addToShoppingList(UUID recipeId) {
        new AddToShoppingListAsyncTask(mDb.getShoppingListDao(), mDb.getRecipeIngredientDao()).execute(recipeId);
    }

    public void update(UiRecipeIngredient... ingredients) {
        new UpdateIngredientAsyncTask(mDb.getRecipeIngredientDao()).execute(ingredients);
    }

    public LiveData<RecipeStep> getRecipeStep(UUID id) {
        return mDb.getRecipeStepDao().findById(id);
    }

    public void updateRecipeStep(RecipeStep recipeStep, OnWebResponseListener listener) {
        new Request<RecipeStep>(mContext) {
            @Override
            public RecipeStep getEntityBeforeUpdate(RecipeStep newEntity) {
                return mDb.getRecipeStepDao().findByIdRaw(newEntity.getId());
            }

            @Override
            public void updateLocalDatabase(RecipeStep entity) {
                mDb.getRecipeStepDao().update(entity);
            }

            @NotNull
            @Override
            public Call<ResourceData<RecipeStep>> sendRequestToServer(RecipeStep newEntity) {
                return mWebservice.updateRecipeStep(newEntity);
            }

            @Override
            public void onSuccess(RecipeStep responseEntity) {
                listener.onSuccess();
            }

            @Override
            public void onError(@NonNull String errorCode, @NonNull String errorMessage) {
                listener.onError(errorCode, errorMessage);
            }
        }.send(recipeStep);
    }

    private static class RemoveFromShoppingListAsyncTask extends AsyncTask<UUID, Void, Void> {

        private ShoppingListDao mAsyncRecipeDao;

        RemoveFromShoppingListAsyncTask(ShoppingListDao dao) {
            mAsyncRecipeDao = dao;
        }

        @Override
        protected Void doInBackground(final UUID... recipeIds) {
            mAsyncRecipeDao.deleteByRecipeId(recipeIds[0]);
            return null;
        }
    }

    private static class AddToShoppingListAsyncTask extends AsyncTask<UUID, Void, Void> {

        private ShoppingListDao shoppingListDao;
        private RecipeIngredientDao recipeIngredientDao;

        AddToShoppingListAsyncTask(ShoppingListDao shoppingListDao, RecipeIngredientDao recipeIngredientDao) {
            this.shoppingListDao = shoppingListDao;
            this.recipeIngredientDao = recipeIngredientDao;
        }

        @Override
        protected Void doInBackground(UUID... recipeIds) {
            Debug.d(this, "Adding recipe with id '" + recipeIds[0] + "' to shopping list");
            final UUID recipeId = recipeIds[0];
            List<ShoppingListItem> shoppingListItems = recipeIngredientDao.findShoppingListItemByRecipeId(recipeId);
            shoppingListDao.deleteByRecipeId(recipeId);
            shoppingListDao.insert(shoppingListItems);
            return null;
        }
    }

    private static class UpdateIngredientAsyncTask extends AsyncTask<UiRecipeIngredient, Void, Void> {

        private RecipeIngredientDao recipeIngredientDao;

        UpdateIngredientAsyncTask(RecipeIngredientDao dao) {
            recipeIngredientDao = dao;
        }

        @Override
        protected Void doInBackground(final UiRecipeIngredient... ingredients) {
            for (UiRecipeIngredient uiRecipeIngredient : ingredients) {
                Debug.d(this, "Saving ingredient " + uiRecipeIngredient.toString());
                RecipeIngredient recipeIngredient = recipeIngredientDao.findByIdRaw(uiRecipeIngredient.getId());
                recipeIngredient.setQuantity(uiRecipeIngredient.getQuantity());
                recipeIngredientDao.update(recipeIngredient);
            }
            return null;
        }
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/database/repository/RecipeDetailRepository.kt
package com.yeraygarcia.recipes.database.repository

import android.app.Application
import android.arch.lifecycle.LiveData
import android.content.Context
import com.yeraygarcia.recipes.AppExecutors
import com.yeraygarcia.recipes.OnWebResponseListener
import com.yeraygarcia.recipes.database.AppDatabase
import com.yeraygarcia.recipes.database.UUIDTypeConverter
import com.yeraygarcia.recipes.database.entity.*
import com.yeraygarcia.recipes.database.entity.Unit
import com.yeraygarcia.recipes.database.entity.custom.UiRecipeIngredient
import com.yeraygarcia.recipes.database.remote.Request
import com.yeraygarcia.recipes.database.remote.ResourceData
import com.yeraygarcia.recipes.database.remote.RetrofitClient
import com.yeraygarcia.recipes.database.remote.Webservice
import com.yeraygarcia.recipes.util.NetworkUtil
import com.yeraygarcia.recipes.util.toJson
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModel
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber
import java.util.*

class RecipeDetailRepository(application: Application) {

    private val context: Context = application
    private val db = AppDatabase.getDatabase(context)
    private val webservice = RetrofitClient.get(context).create(Webservice::class.java)
    private val appExecutors = AppExecutors()

    val units: LiveData<List<Unit>>
        get() = db.unitDao.findAll()

    val ingredientNames: LiveData<List<String>>
        get() = db.ingredientDao.ingredientNames

    val unitNames: LiveData<List<String>>
        get() = db.unitDao.unitNames

    fun getRecipeById(id: UUID): LiveData<Recipe> {
        return db.recipeDao.findById(id)
    }

    fun getIngredientsByRecipeId(recipeId: UUID): LiveData<List<UiRecipeIngredient>> {
        return db.recipeIngredientDao.findByRecipeId(recipeId)
    }

    fun getStepsByRecipeId(recipeId: UUID): LiveData<List<RecipeStep>> {
        return db.recipeStepDao.findByRecipeId(recipeId)
    }

    fun update(recipe: Recipe) {
        object : Request<Recipe>(context) {
            override fun getEntityBeforeUpdate(newEntity: Recipe): Recipe {
                return db.recipeDao.findRawById(newEntity.id)
            }

            override fun updateLocalDatabase(entity: Recipe) {
                db.recipeDao.update(entity)
            }

            override fun sendRequestToServer(newEntity: Recipe): Call<ResourceData<Recipe>> {
                return webservice.updateRecipe(newEntity)
            }

            override fun onSuccess(responseEntity: Recipe) {
                // do nothing
            }

            override fun onError(errorCode: String, errorMessage: String) {
                // do nothing
            }
        }.send(recipe)
    }

    fun persistDraft(recipe: Recipe, recipeDetailViewModel: RecipeDetailViewModel) {
        object : Request<Recipe>(context) {
            override fun getEntityBeforeUpdate(newEntity: Recipe): Recipe {
                return db.recipeDao.findRawById(newEntity.id)
            }

            override fun updateLocalDatabase(entity: Recipe) {
                db.recipeDao.update(entity)
            }

            override fun sendRequestToServer(newEntity: Recipe): Call<ResourceData<Recipe>> {
                return webservice.updateRecipe(newEntity)
            }

            override fun onSuccess(responseEntity: Recipe) {
                recipeDetailViewModel.recipeDraft = null
            }

            override fun onError(errorCode: String, errorMessage: String) {
                // do nothing
            }
        }.send(recipe)
    }

    fun isInShoppingList(recipeId: UUID): LiveData<Boolean> {
        return db.shoppingListDao.isInShoppingList(recipeId)
    }

    fun removeFromShoppingList(recipeId: UUID) {
        appExecutors.diskIO { db.shoppingListDao.deleteByRecipeId(recipeId) }
    }

    fun addToShoppingList(recipeId: UUID) {
        appExecutors.diskIO {
            Timber.d("Adding recipe with id '$recipeId' to shopping list")
            val shoppingListItems = db.recipeIngredientDao.findShoppingListItemByRecipeId(recipeId)
            shoppingListItems.forEach { it.id = UUIDTypeConverter.newUUID() }
            db.shoppingListDao.deleteByRecipeId(recipeId)
            db.shoppingListDao.insert(shoppingListItems)
        }
    }

    fun getRecipeStep(id: UUID): LiveData<RecipeStep> {
        return db.recipeStepDao.findById(id)
    }

    fun updateRecipeStep(recipeStep: RecipeStep, listener: OnWebResponseListener) {
        object : Request<RecipeStep>(context) {
            override fun getEntityBeforeUpdate(newEntity: RecipeStep): RecipeStep {
                return db.recipeStepDao.findByIdRaw(newEntity.id)
            }

            override fun updateLocalDatabase(entity: RecipeStep) {
                db.recipeStepDao.update(entity)
            }

            override fun sendRequestToServer(newEntity: RecipeStep): Call<ResourceData<RecipeStep>> {
                return webservice.updateRecipeStep(newEntity)
            }

            override fun onSuccess(responseEntity: RecipeStep) {
                listener.onSuccess()
            }

            override fun onError(errorCode: String, errorMessage: String) {
                listener.onError(errorCode, errorMessage)
            }
        }.send(recipeStep)
    }

    fun getUiRecipeIngredient(id: UUID): LiveData<UiRecipeIngredient> {
        return db.recipeIngredientDao.findById(id)
    }

    private fun canFetch(): Boolean {
        return NetworkUtil.isOnline(context)
    }

    fun updateRecipeIngredient(id: UUID, ingredientName: String, quantity: Double?, unitId: UUID?) {
        Timber.d("updateRecipeIngredient(id, ingredientName, quantity, unitId)")
        if (canFetch()) {
            appExecutors.diskIO {
                var ingredientId: UUID? = db.ingredientDao.getIngredientIdByName(ingredientName)
                if (ingredientId == null) {
                    val ingredient = Ingredient(name = ingredientName)
                    ingredientId = ingredient.id
                    db.ingredientDao.insert(ingredient)
                }
                val oldRecipeIngredient = db.recipeIngredientDao.findByIdRaw(id)
                val newRecipeIngredient = oldRecipeIngredient.copy(
                    ingredientId = ingredientId,
                    quantity = quantity,
                    unitId = unitId
                )

                db.recipeIngredientDao.update(newRecipeIngredient)

                Timber.d("Sending request to server.")
                Timber.d(newRecipeIngredient.toString())

                val call = webservice.updateRecipeIngredient(newRecipeIngredient)

                call.enqueue(object : Callback<ResourceData<RecipeIngredient>> {
                    override fun onResponse(
                        call: Call<ResourceData<RecipeIngredient>>,
                        response: Response<ResourceData<RecipeIngredient>>
                    ) {
                        Timber.d("We got a response from the server: ${response.toJson()}")
                        if (response.isSuccessful) {
                            Timber.d("Response was successful: ${response.code()}")
                            val body = response.body()
                            if (body?.result != null) {
                                Timber.d("Body contains payload. Update database.")
                            } else {
                                Timber.d("Body was empty. Abort.")
                                appExecutors.diskIO {
                                    db.recipeIngredientDao.update(
                                        oldRecipeIngredient
                                    )
                                }
                            }
                        } else {
                            Timber.d("Response was not successful: ${response.code()} - ${response.errorBody()}")
                            appExecutors.diskIO { db.recipeIngredientDao.update(oldRecipeIngredient) }
                        }
                    }

                    override fun onFailure(
                        call: Call<ResourceData<RecipeIngredient>>,
                        t: Throwable
                    ) {
                        Timber.d("Call failed: ${t.message}")
                        appExecutors.diskIO {
                            db.recipeIngredientDao.update(oldRecipeIngredient)
                        }
                    }
                })
            }
        }
    }

}
